src/java.base/share/classes/sun/security/ssl/ServerHello.java

```diff
@@ -590,9 +616,26 @@ public byte[] produce(ConnectionContext context,
                         "Not negotiated key shares");
             }
 
-            SSLKeyDerivation handshakeKD = ke.createKeyDerivation(shc);
-            SecretKey handshakeSecret = handshakeKD.deriveKey(
-                    "TlsHandshakeSecret");
+            SecretKey handshakeSecret = null;
+
+            // For KEM, the shared secret has already been generated and
+            // stored in the server’s possession (KEMSenderPossession)
+            // during encapsulation in SHKeyShareProducer.
+            //
+            // Only one key share is selected by the server, so at most one
+            // possession will contain the pre-derived shared secret.
+            for (var pos : shc.handshakePossessions) {
+                if (pos instanceof KEMKeyExchange.KEMSenderPossession xp) {
+                    handshakeSecret = xp.getKey();
+                    break;
+                }
+            }
+
+            if (handshakeSecret == null) {
+                SSLKeyDerivation handshakeKD = ke.createKeyDerivation(shc);
```

Comment:
Since we only end up with one possession, can we move this up into the for loop?  Each of the various possession types goes through the `for` loop, so seems this would be a tiny bit clearer.

    for () {
        if (pos instanceof ...) {
            handshakeSecret =...
        } else {
            SSLKeyDerivation handshakeKD = ...
        }
    }

Reply:
Keep the code as it is. This is because if we move the KA derivation code to the for loop as the else branch, we make it appear to be associated with a possession. KA does not store the shared secret in a possession like KEM, so keep its logic separate.